src/modules/tools/jogger/Jogger.cpp

```diff
@@ -0,0 +1,292 @@
+        }
+    }
+}
+
+//read config file values for this module
+void Jogger::on_config_reload(void *argument)
+{
+    //load the basic configuration settings
+    this->max_speed = THEKERNEL->config->value(default_seek_rate_checksum)->by_default(this->max_speed)->as_number();
+    this->max_speed = THEKERNEL->config->value(jogger_checksum, max_speed_checksum)->by_default(this->max_speed)->as_number();
+    this->dead_zone = THEKERNEL->config->value(jogger_checksum, dead_zone_checksum)->by_default(this->dead_zone)->as_number();
+    this->nonlinearity = THEKERNEL->config->value(jogger_checksum, nonlinearity_checksum)->by_default(this->nonlinearity)->as_number();
+    this->refresh_rate = THEKERNEL->config->value(jogger_checksum, refresh_rate_checksum)->by_default(this->refresh_rate)->as_number();
+    this->segment_frequency = THEKERNEL->config->value(jogger_checksum, segment_frequency_checksum)->by_default(this->segment_frequency)->as_number();
+    this->m_code_set = THEKERNEL->config->value(jogger_checksum, m_code_set_checksum)->by_default(this->m_code_set)->as_number();
+    this->m_code_toggle = THEKERNEL->config->value(jogger_checksum, m_code_toggle_checksum)->by_default(this->m_code_toggle)->as_number();
+    
+    //load the names of the joystick modules where each axis will get its data
+    uint16_t axisN_data_source_checksum[] = { axis0_data_source_checksum, axis1_data_source_checksum, axis2_data_source_checksum, axis3_data_source_checksum, axis4_data_source_checksum, axis5_data_source_checksum };
+    for (int i = 0; i < NUM_JOG_AXES; i++) {
+        this->axis_data_source[i] = get_checksum(THEKERNEL->config->value(jogger_checksum, axisN_data_source_checksum[i])->by_default("")->as_string());
+    }
+
+    //load the machine axis letters that the joystick will control
+    std::string letterlist = THEKERNEL->config->value(jogger_checksum, jog_axes_checksum)->by_default("XYZABC")->as_string();
+
+    //parse the letter list into substrings
+    this->jog_axes = split(letterlist.c_str(), ',');
+
+    //update the current machine axis settings
+    update_Axes(this->jog_axes[this->axis_index]);
+}
+
+//map a position from -1 to 1 into a speed from -max to max
+float Jogger::get_speed(float pos) {
+    float speed;
+
+    //zero out in dead zone
+    if (abs(pos) < this->dead_zone) {
+        speed = 0.0f;
+    }
+    else {
+        //apply non-linear scaling to position
+        speed = (sign(pos) * pow(abs(pos), this->nonlinearity)) * this->max_speed;
+    }
+
+    return speed;
+}
+
+//runs on a timer to update the jog speeds (default 100 Hz update rate)
+uint32_t Jogger::update_tick(uint32_t dummy)
+{
+    bool allzero = true; //will determine whether the joystick is at rest or not
+
+    //read the joysticks for each axis and map their values to a target speed
+    for (int c = 0; c < NUM_JOG_AXES; c++) {
+        struct PAD_joystick s;
+        if (PublicData::get_value(joystick_checksum, this->axis_data_source[c], &s)) {
```

Comment:
Is this something that can be moved to the on_main_loop? Ideally I want the jogger and joystick modules to stay separate, so that jogging can be done with any type of device (e.g. encoder) and the joystick can be used in any other module (e.g. feedrate override)